// @sap-ux/edmx-parser/dist/v2annotationsSupport.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.convertEntityTypeAnnotations = exports.convertPropertyAnnotations = exports.convertNavigationPropertyAnnotations = exports.convertEntitySetAnnotations = exports.convertGenericAnnotations = exports.convertV2Annotations = void 0;
/**
 * Convert v2 annotation that were defined on the schema as standard v4 annotations.
 *
 * @param attributes the attribute of the current object
 * @param objectType the object type
 * @param objectName the object name
 * @returns the converted annotations
 */
function convertV2Annotations(attributes, objectType, objectName) {
    const annotations = [];
    switch (objectType) {
        case 'EntitySet':
            convertEntitySetAnnotations(attributes, annotations);
            break;
        case 'EntityType':
            convertEntityTypeAnnotations(attributes, annotations);
            break;
        case 'NavigationProperty':
            convertNavigationPropertyAnnotations(attributes, objectName, annotations);
            break;
        case 'Property':
            convertPropertyAnnotations(attributes, objectName, annotations);
            break;
        case 'Singleton':
            break;
        default:
            break;
    }
    convertGenericAnnotations(attributes, objectName, annotations);
    return annotations;
}
exports.convertV2Annotations = convertV2Annotations;
/**
 * Convert annotation that can apply to all kind of objects.
 *
 * @param attributes the attribute of the current object
 * @param objectName the object name
 * @param annotations the raw annotation array
 */
function convertGenericAnnotations(attributes, objectName, annotations) {
    /**
     * Push to annotation if the condition evaluates to true.
     *
     * @param condition the condition to evaluate
     * @param value the target value
     */
    function pushToAnnotations(condition, value) {
        if (condition) {
            annotations.push(value);
        }
    }
    pushToAnnotations(attributes['sap:schema-version'] !== undefined, {
        term: "Org.OData.Core.V1.SchemaVersion" /* CoreAnnotationTerms.SchemaVersion */,
        value: {
            type: 'String',
            String: attributes['sap:schema-version']
        }
    });
    pushToAnnotations(attributes['sap:searchable'] !== undefined, {
        term: "Org.OData.Capabilities.V1.SearchRestrictions" /* CapabilitiesAnnotationTerms.SearchRestrictions */,
        record: {
            propertyValues: [
                {
                    name: 'Searchable',
                    value: {
                        type: 'Bool',
                        Bool: attributes['sap:searchable'] === 'true'
                    }
                }
            ]
        }
    });
    pushToAnnotations(attributes['sap:pageable'] !== undefined, {
        term: "Org.OData.Capabilities.V1.TopSupported" /* CapabilitiesAnnotationTerms.TopSupported */,
        value: {
            type: 'Bool',
            Bool: attributes['sap:pageable'] === 'true'
        }
    });
    pushToAnnotations(attributes['sap:pageable'] !== undefined, {
        term: "Org.OData.Capabilities.V1.SkipSupported" /* CapabilitiesAnnotationTerms.SkipSupported */,
        value: {
            type: 'Bool',
            Bool: attributes['sap:pageable'] === 'true'
        }
    });
    pushToAnnotations(attributes['sap:topable'] !== undefined, {
        term: "Org.OData.Capabilities.V1.TopSupported" /* CapabilitiesAnnotationTerms.TopSupported */,
        value: {
            type: 'Bool',
            Bool: attributes['sap:topable'] === 'true'
        }
    });
    pushToAnnotations(attributes['sap:requires-filter'] !== undefined, {
        term: "Org.OData.Capabilities.V1.FilterRestrictions" /* CapabilitiesAnnotationTerms.FilterRestrictions */,
        record: {
            propertyValues: [
                {
                    name: 'RequiresFilter',
                    value: {
                        type: 'Bool',
                        Bool: attributes['sap:requires-filter'] === 'true'
                    }
                }
            ]
        }
    });
    pushToAnnotations(attributes['sap:required-in-filter'] !== undefined, {
        term: "Org.OData.Capabilities.V1.FilterRestrictions" /* CapabilitiesAnnotationTerms.FilterRestrictions */,
        record: {
            propertyValues: [
                {
                    name: 'RequiredProperties',
                    value: {
                        type: 'Collection',
                        Collection: [
                            {
                                type: 'PropertyPath',
                                PropertyPath: objectName
                            }
                        ]
                    }
                }
            ]
        }
    });
    pushToAnnotations(attributes['sap:filter-restricton'] !== undefined, {
        term: "Org.OData.Capabilities.V1.FilterRestrictions" /* CapabilitiesAnnotationTerms.FilterRestrictions */,
        record: {
            propertyValues: [
                {
                    name: 'FilterExpressionRestrictions',
                    value: {
                        type: 'Collection',
                        Collection: [
                            {
                                type: 'Record',
                                propertyValues: [
                                    {
                                        name: 'FilterExpressionRestrictions',
                                        value: {
                                            type: 'String',
                                            String: attributes['sap:filter-restricton']
                                        }
                                    },
                                    {
                                        name: 'Property',
                                        value: {
                                            type: 'PropertyPath',
                                            PropertyPath: objectName
                                        }
                                    }
                                ]
                            }
                        ]
                    }
                }
            ]
        }
    });
    pushToAnnotations(attributes['sap:sortable'] === 'false', {
        term: "Org.OData.Capabilities.V1.SortRestrictions" /* CapabilitiesAnnotationTerms.SortRestrictions */,
        record: {
            propertyValues: [
                {
                    name: 'NonSortableProperties',
                    value: {
                        type: 'PropertyPath',
                        PropertyPath: objectName
                    }
                }
            ]
        }
    });
    pushToAnnotations(attributes['sap:visible'] === 'false', {
        term: "com.sap.vocabularies.UI.v1.Hidden" /* UIAnnotationTerms.Hidden */,
        value: {
            type: 'Bool',
            Bool: true
        }
    });
    pushToAnnotations(attributes['sap:label'] !== undefined, {
        term: "com.sap.vocabularies.Common.v1.Label" /* CommonAnnotationTerms.Label */,
        value: {
            type: 'String',
            String: attributes['sap:label']
        }
    });
    pushToAnnotations(attributes['sap:heading'] !== undefined, {
        term: "com.sap.vocabularies.Common.v1.Heading" /* CommonAnnotationTerms.Heading */,
        value: {
            type: 'String',
            String: attributes['sap:heading']
        }
    });
    pushToAnnotations(attributes['sap:quickinfo'] !== undefined, {
        term: "com.sap.vocabularies.Common.v1.QuickInfo" /* CommonAnnotationTerms.QuickInfo */,
        value: {
            type: 'String',
            String: attributes['sap:quickinfo']
        }
    });
    pushToAnnotations(attributes['sap:text'] !== undefined, {
        term: "com.sap.vocabularies.Common.v1.Text" /* CommonAnnotationTerms.Text */,
        value: {
            type: 'Path',
            Path: attributes['sap:text']
        }
    });
    pushToAnnotations(attributes['sap:unit'] !== undefined, {
        term: "Org.OData.Measures.V1.Unit" /* MeasuresAnnotationTerms.Unit */,
        value: {
            type: 'Path',
            Path: attributes['sap:unit']
        }
    });
    pushToAnnotations(attributes['sap:unit'] !== undefined, {
        term: "Org.OData.Measures.V1.ISOCurrency" /* MeasuresAnnotationTerms.ISOCurrency */,
        value: {
            type: 'Path',
            Path: attributes['sap:unit']
        }
    });
    pushToAnnotations(attributes['sap:precision'] !== undefined, {
        term: "Org.OData.Measures.V1.Scale" /* MeasuresAnnotationTerms.Scale */,
        value: {
            type: 'Int',
            Int: parseInt(attributes['sap:precision'], 10)
        }
    });
    pushToAnnotations(attributes['sap:value-list'] === 'fixed-value', {
        term: "com.sap.vocabularies.Common.v1.ValueListWithFixedValues" /* CommonAnnotationTerms.ValueListWithFixedValues */,
        value: {
            type: 'Bool',
            Bool: true
        }
    });
    pushToAnnotations(attributes['sap:display-format'] === 'NonNegative', {
        term: "com.sap.vocabularies.Common.v1.IsDigitSequence" /* CommonAnnotationTerms.IsDigitSequence */,
        value: {
            type: 'Bool',
            Bool: true
        }
    });
    pushToAnnotations(attributes['sap:display-format'] === 'UpperCase', {
        term: "com.sap.vocabularies.Common.v1.IsUpperCase" /* CommonAnnotationTerms.IsUpperCase */,
        value: {
            type: 'Bool',
            Bool: true
        }
    });
    if (attributes['sap:lower-boundary'] || attributes['sap:upper-boundary']) {
        const pv = [];
        if (attributes['sap:lower-boundary']) {
            pv.push({
                name: 'LowerBoundary',
                value: {
                    type: 'PropertyPath',
                    PropertyPath: attributes['sap:lower-boundary']
                }
            });
        }
        if (attributes['sap:upper-boundary']) {
            pv.push({
                name: 'UpperBoundary',
                value: {
                    type: 'PropertyPath',
                    PropertyPath: attributes['sap:upper-boundary']
                }
            });
        }
        annotations.push({
            term: "com.sap.vocabularies.Common.v1.Interval" /* CommonAnnotationTerms.Interval */,
            record: {
                propertyValues: pv
            }
        });
    }
    pushToAnnotations(attributes['sap:field-control'] !== undefined, {
        term: "com.sap.vocabularies.Common.v1.FieldControl" /* CommonAnnotationTerms.FieldControl */,
        value: {
            type: 'Path',
            Path: attributes['sap:field-control']
        }
    });
    pushToAnnotations(attributes['sap:applicable-path'] !== undefined, {
        term: "Org.OData.Core.V1.OperationAvailable" /* CoreAnnotationTerms.OperationAvailable */,
        value: {
            type: 'Path',
            Path: attributes['sap:applicable-path']
        }
    });
    pushToAnnotations(attributes['sap:minoccurs'] !== undefined, {
        term: "com.sap.vocabularies.Common.v1.MinOccurs" /* CommonAnnotationTerms.MinOccurs */,
        value: {
            type: 'Int',
            Int: parseInt(attributes['sap:minoccurs'], 10)
        }
    });
    pushToAnnotations(attributes['sap:maxoccurs'] !== undefined, {
        term: "com.sap.vocabularies.Common.v1.MaxOccurs" /* CommonAnnotationTerms.MaxOccurs */,
        value: {
            type: 'Int',
            Int: parseInt(attributes['sap:maxoccurs'], 10)
        }
    });
    pushToAnnotations(attributes['sap:parameter'] === 'mandatory', {
        term: "com.sap.vocabularies.Common.v1.FieldControl" /* CommonAnnotationTerms.FieldControl */,
        value: {
            type: 'EnumMember',
            EnumMember: "Common.FieldControlType/Mandatory" /* FieldControlType.Mandatory */
        }
    });
    pushToAnnotations(attributes['sap:parameter'] === 'optional', {
        term: "com.sap.vocabularies.Common.v1.FieldControl" /* CommonAnnotationTerms.FieldControl */,
        value: {
            type: 'EnumMember',
            EnumMember: "Common.FieldControlType/Optional" /* FieldControlType.Optional */
        }
    });
    pushToAnnotations(attributes['sap:attribute-for'] !== undefined, {
        term: "com.sap.vocabularies.Common.v1.Attributes" /* CommonAnnotationTerms.Attributes */,
        value: {
            type: 'Collection',
            Collection: [
                {
                    type: 'PropertyPath',
                    PropertyPath: objectName
                }
            ]
        }
    });
}
exports.convertGenericAnnotations = convertGenericAnnotations;
/**
 * Convert annotations specific to entityset.
 *
 * @param attributes The V2 Annotations to evaluate
 * @param annotations The raw annotation array
 */
function convertEntitySetAnnotations(attributes, annotations) {
    if (attributes['sap:creatable']) {
        annotations.push({
            term: "Org.OData.Capabilities.V1.InsertRestrictions" /* CapabilitiesAnnotationTerms.InsertRestrictions */,
            record: {
                propertyValues: [
                    {
                        name: 'Insertable',
                        value: {
                            type: 'Bool',
                            Bool: attributes['sap:creatable'] === 'true'
                        }
                    }
                ]
            }
        });
    }
    if (attributes['sap:updatable']) {
        annotations.push({
            term: "Org.OData.Capabilities.V1.UpdateRestrictions" /* CapabilitiesAnnotationTerms.UpdateRestrictions */,
            record: {
                propertyValues: [
                    {
                        name: 'Updatable',
                        value: {
                            type: 'Bool',
                            Bool: attributes['sap:updatable'] === 'true'
                        }
                    }
                ]
            }
        });
    }
    if (attributes['sap:updatable-path']) {
        annotations.push({
            term: "Org.OData.Capabilities.V1.UpdateRestrictions" /* CapabilitiesAnnotationTerms.UpdateRestrictions */,
            record: {
                propertyValues: [
                    {
                        name: 'Updatable',
                        value: {
                            type: 'Path',
                            Path: attributes['sap:updatable-path']
                        }
                    }
                ]
            }
        });
    }
    if (attributes['sap:deletable']) {
        annotations.push({
            term: "Org.OData.Capabilities.V1.DeleteRestrictions" /* CapabilitiesAnnotationTerms.DeleteRestrictions */,
            record: {
                propertyValues: [
                    {
                        name: 'Deletable',
                        value: {
                            type: 'Bool',
                            Bool: attributes['sap:updatable'] === 'true'
                        }
                    }
                ]
            }
        });
    }
    if (attributes['sap:deletable-path']) {
        annotations.push({
            term: "Org.OData.Capabilities.V1.DeleteRestrictions" /* CapabilitiesAnnotationTerms.DeleteRestrictions */,
            record: {
                propertyValues: [
                    {
                        name: 'Deletable',
                        value: {
                            type: 'Path',
                            Path: attributes['sap:deletable-path']
                        }
                    }
                ]
            }
        });
    }
}
exports.convertEntitySetAnnotations = convertEntitySetAnnotations;
/**
 * Convert annotations specific to navigation properties.
 *
 * @param attributes The V2 Annotations to evaluate
 * @param objectName The name of the navigation property
 * @param annotations The raw annotation array
 */
function convertNavigationPropertyAnnotations(attributes, objectName, annotations) {
    if (attributes['sap:creatable']) {
        annotations.push({
            term: "Org.OData.Capabilities.V1.NavigationRestrictions" /* CapabilitiesAnnotationTerms.NavigationRestrictions */,
            record: {
                propertyValues: [
                    {
                        name: 'RestrictedProperties',
                        value: {
                            type: 'Record',
                            Record: {
                                propertyValues: [
                                    {
                                        name: 'InsertRestrictrions',
                                        value: {
                                            type: 'Record',
                                            Record: {
                                                propertyValues: [
                                                    {
                                                        name: 'Insertable',
                                                        value: {
                                                            type: 'Bool',
                                                            Bool: attributes['sap:creatable'] === 'true'
                                                        }
                                                    }
                                                ]
                                            }
                                        }
                                    }
                                ]
                            }
                        }
                    }
                ]
            }
        });
    }
    if (attributes['sap:creatable-path']) {
        annotations.push({
            term: "Org.OData.Capabilities.V1.NavigationRestrictions" /* CapabilitiesAnnotationTerms.NavigationRestrictions */,
            record: {
                propertyValues: [
                    {
                        name: 'RestrictedProperties',
                        value: {
                            type: 'Record',
                            Record: {
                                propertyValues: [
                                    {
                                        name: 'InsertRestrictrions',
                                        value: {
                                            type: 'Record',
                                            Record: {
                                                propertyValues: [
                                                    {
                                                        name: 'Insertable',
                                                        value: {
                                                            type: 'Path',
                                                            Path: attributes['sap:creatable-path']
                                                        }
                                                    }
                                                ]
                                            }
                                        }
                                    }
                                ]
                            }
                        }
                    }
                ]
            }
        });
    }
    if (attributes['sap:filterable'] === 'false') {
        annotations.push({
            term: "Org.OData.Capabilities.V1.NavigationRestrictions" /* CapabilitiesAnnotationTerms.NavigationRestrictions */,
            record: {
                propertyValues: [
                    {
                        name: 'RestrictedProperties',
                        value: {
                            type: 'Collection',
                            Collection: [
                                {
                                    type: 'Record',
                                    propertyValues: [
                                        {
                                            name: 'NavigationProperty',
                                            value: {
                                                type: 'NavigationPropertyPath',
                                                NavigationPropertyPath: objectName
                                            }
                                        },
                                        {
                                            name: 'FilterRestrictions',
                                            value: {
                                                type: 'Record',
                                                Record: {
                                                    propertyValues: [
                                                        {
                                                            name: 'Filterable',
                                                            value: {
                                                                type: 'Bool',
                                                                Bool: false
                                                            }
                                                        }
                                                    ]
                                                }
                                            }
                                        }
                                    ]
                                }
                            ]
                        }
                    }
                ]
            }
        });
    }
}
exports.convertNavigationPropertyAnnotations = convertNavigationPropertyAnnotations;
/**
 * Convert annotations specific to properties.
 *
 * @param attributes The V2 Annotations to evaluate
 * @param objectName The name of the property
 * @param annotations The raw annotation array
 */
function convertPropertyAnnotations(attributes, objectName, annotations) {
    if (attributes['sap:creatable'] === 'true' && attributes['sap:updatable'] === 'false') {
        annotations.push({
            term: "Org.OData.Core.V1.Immutable" /* CoreAnnotationTerms.Immutable */,
            value: {
                type: 'Bool',
                Bool: true
            }
        });
    }
    if (attributes['sap:creatable'] === 'false' && attributes['sap:updatable'] === 'false') {
        annotations.push({
            term: "Org.OData.Core.V1.Computed" /* CoreAnnotationTerms.Computed */,
            value: {
                type: 'Bool',
                Bool: true
            }
        });
    }
    if (attributes['sap:updatable-path']) {
        annotations.push({
            term: "com.sap.vocabularies.Common.v1.FieldControl" /* CommonAnnotationTerms.FieldControl */,
            value: {
                type: 'Path',
                Path: attributes['sap:updatable-path']
            }
        });
    }
    if (attributes['sap:filterable'] === 'false') {
        annotations.push({
            term: "Org.OData.Capabilities.V1.FilterRestrictions" /* CapabilitiesAnnotationTerms.FilterRestrictions */,
            record: {
                propertyValues: [
                    {
                        name: 'NonFilterableProperties',
                        value: {
                            type: 'Collection',
                            Collection: [
                                {
                                    type: 'PropertyPath',
                                    PropertyPath: objectName
                                }
                            ]
                        }
                    }
                ]
            }
        });
    }
}
exports.convertPropertyAnnotations = convertPropertyAnnotations;
/**
 * Convert annotations specific to properties.
 *
 * @param attributes The V2 Annotations to evaluate
 * @param annotations The raw annotation array
 */
function convertEntityTypeAnnotations(attributes, annotations) {
    if (attributes['sap:semantics'] === 'aggregate') {
        annotations.push({
            term: "Org.OData.Aggregation.V1.ApplySupported" /* AggregationAnnotationTerms.ApplySupported */,
            record: {
                propertyValues: [
                    {
                        /**
                         Only properties marked as `Groupable` can be used in the `groupby` transformation, and only those marked as `Aggregatable` can be used in the  `aggregate` transformation
                         */
                        name: 'PropertyRestrictions',
                        value: {
                            type: 'Collection',
                            Collection: []
                        }
                    },
                    {
                        /**
                         A non-empty collection indicates that only the listed properties of the annotated target are supported by the `groupby` transformation
                         */
                        name: 'GroupableProperties',
                        value: {
                            type: 'Collection',
                            Collection: []
                        }
                    },
                    {
                        /**
                         A non-empty collection indicates that only the listed properties of the annotated target can be used in the `aggregate` transformation, optionally restricted to the specified aggregation methods
                         */
                        name: 'AggregatableProperties',
                        value: {
                            type: 'Collection',
                            Collection: []
                        }
                    }
                ]
            }
        });
    }
}
exports.convertEntityTypeAnnotations = convertEntityTypeAnnotations;
