// @sap/ux-specification/dist/src/sync/v2/export/controls/ObjectPageSectionsV2.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.ObjectPageSubSectionTableV2 = exports.ObjectPageSubSectionContactV2 = exports.ObjectPageSubSectionAddressV2 = exports.ObjectPageSubSectionChartV2 = exports.ObjectPageSubSectionFormV2 = exports.ObjectPageSubSectionV2 = exports.ObjectPageSectionContactV2 = exports.ObjectPageSectionAddressV2 = exports.ObjectPageSectionTableV2 = exports.ObjectPageSectionChartV2 = exports.ObjectPageSectionFormV2 = exports.ObjectPageSectionV2 = exports.getSubSectionControlId = exports.getSectionControlId = exports.getChartSectionControlId = exports.getFormSectionControlId = exports.getTableSectionControlId = void 0;
const ux_specification_types_1 = require("@sap/ux-specification-types");
const decorators_1 = require("../../../common/decoration/decorators");
const common_1 = require("../../../common");
const getTableSectionControlId = (baseId, _idBreadcrumbs, breadcrumbs, _controlType, facetId) => {
    const convertedId = (0, common_1.convertSectionId)(breadcrumbs[1], facetId);
    const breadcrumbsTrail = [convertedId, ux_specification_types_1.DefinitionName.Table];
    return (0, common_1.buildControlIdFromParent)(baseId, breadcrumbsTrail);
};
exports.getTableSectionControlId = getTableSectionControlId;
const getFormSectionControlId = (baseId, _idBreadcrumbs, breadcrumbs, _controlType, facetId) => {
    const convertedId = (0, common_1.convertSectionId)(breadcrumbs[1], facetId);
    const breadcrumbsTrail = [convertedId, 'FormGroup'];
    return (0, common_1.buildControlIdFromParent)(baseId, breadcrumbsTrail);
};
exports.getFormSectionControlId = getFormSectionControlId;
const getChartSectionControlId = (baseId, _idBreadcrumbs, breadcrumbs, _controlType, facetID) => {
    const convertedId = (0, common_1.convertSectionId)(breadcrumbs[1], facetID);
    const breadcrumbsTrail = [convertedId, 'Chart'];
    return (0, common_1.buildControlIdFromParent)(baseId, breadcrumbsTrail);
};
exports.getChartSectionControlId = getChartSectionControlId;
const getSectionControlId = (baseId, _idBreadcrumbs, breadcrumbs, _controlType, facetId) => {
    const convertedId = (0, common_1.convertSectionId)(breadcrumbs[1], facetId);
    const breadcrumbsTrail = [convertedId, 'Section'];
    return (0, common_1.buildControlIdFromParent)(baseId, breadcrumbsTrail);
};
exports.getSectionControlId = getSectionControlId;
const getSubSectionControlId = (baseId, _idBreadcrumbs, breadcrumbs, _controlType, facetId) => {
    const convertedId = (0, common_1.convertSectionId)(breadcrumbs[1], facetId);
    const breadcrumbsTrail = [convertedId, 'SubSection'];
    return (0, common_1.buildControlIdFromParent)(baseId, breadcrumbsTrail);
};
exports.getSubSectionControlId = getSubSectionControlId;
class ObjectPageSectionV2 {
}
__decorate([
    (0, decorators_1.syncRule)({
        flex: {
            controlId: exports.getSectionControlId,
            controlType: () => ux_specification_types_1.ControlType.Section,
            createFlexChange: common_1.exportToFlexChange
        },
        generate: common_1.addPatternForBindingChangeOfEnumOP
    })
], ObjectPageSectionV2.prototype, "visible", void 0);
exports.ObjectPageSectionV2 = ObjectPageSectionV2;
class ObjectPageSectionFormV2 {
}
__decorate([
    (0, decorators_1.syncRule)({
        flex: {
            controlId: exports.getSectionControlId,
            controlType: () => ux_specification_types_1.ControlType.Section,
            createFlexChange: common_1.exportToFlexChange
        },
        generate: common_1.addPatternForBindingChangeOfEnumOP
    })
], ObjectPageSectionFormV2.prototype, "visible", void 0);
exports.ObjectPageSectionFormV2 = ObjectPageSectionFormV2;
class ObjectPageSectionChartV2 {
}
__decorate([
    (0, decorators_1.syncRule)({
        flex: {
            controlId: exports.getSectionControlId,
            controlType: () => ux_specification_types_1.ControlType.Section,
            createFlexChange: common_1.exportToFlexChange
        },
        generate: common_1.addPatternForBindingChangeOfEnumOP
    })
], ObjectPageSectionChartV2.prototype, "visible", void 0);
exports.ObjectPageSectionChartV2 = ObjectPageSectionChartV2;
class ObjectPageSectionTableV2 {
}
__decorate([
    (0, decorators_1.syncRule)({
        flex: {
            controlId: exports.getSectionControlId,
            controlType: () => ux_specification_types_1.ControlType.Section,
            createFlexChange: common_1.exportToFlexChange
        },
        generate: common_1.addPatternForBindingChangeOfEnumOP
    })
], ObjectPageSectionTableV2.prototype, "visible", void 0);
exports.ObjectPageSectionTableV2 = ObjectPageSectionTableV2;
class ObjectPageSectionAddressV2 {
}
__decorate([
    (0, decorators_1.syncRule)({
        flex: {
            controlId: exports.getSectionControlId,
            controlType: () => ux_specification_types_1.ControlType.Section,
            createFlexChange: common_1.exportToFlexChange
        },
        generate: common_1.addPatternForBindingChangeOfEnumOP
    })
], ObjectPageSectionAddressV2.prototype, "visible", void 0);
exports.ObjectPageSectionAddressV2 = ObjectPageSectionAddressV2;
class ObjectPageSectionContactV2 {
}
__decorate([
    (0, decorators_1.syncRule)({
        flex: {
            controlId: exports.getSectionControlId,
            controlType: () => ux_specification_types_1.ControlType.Section,
            createFlexChange: common_1.exportToFlexChange
        },
        generate: common_1.addPatternForBindingChangeOfEnumOP
    })
], ObjectPageSectionContactV2.prototype, "visible", void 0);
exports.ObjectPageSectionContactV2 = ObjectPageSectionContactV2;
class ObjectPageSubSectionV2 {
}
__decorate([
    (0, decorators_1.syncRule)({
        flex: {
            controlId: exports.getSubSectionControlId,
            controlType: () => ux_specification_types_1.ControlType.SubSection,
            createFlexChange: common_1.exportToFlexChange
        },
        generate: common_1.addPatternForBindingChangeOfEnumOP
    })
], ObjectPageSubSectionV2.prototype, "visible", void 0);
exports.ObjectPageSubSectionV2 = ObjectPageSubSectionV2;
class ObjectPageSubSectionFormV2 {
}
__decorate([
    (0, decorators_1.syncRule)({
        flex: {
            controlId: exports.getFormSectionControlId,
            controlType: () => ux_specification_types_1.ControlType.Group,
            createFlexChange: common_1.exportToFlexChange
        },
        generate: common_1.addPatternForBindingChangeOfEnumOP
    })
], ObjectPageSubSectionFormV2.prototype, "visible", void 0);
exports.ObjectPageSubSectionFormV2 = ObjectPageSubSectionFormV2;
class ObjectPageSubSectionChartV2 {
}
__decorate([
    (0, decorators_1.syncRule)({
        flex: {
            controlId: exports.getChartSectionControlId,
            controlType: () => ux_specification_types_1.ControlType.SmartChart,
            createFlexChange: common_1.exportToFlexChange
        },
        generate: common_1.addPatternForBindingChangeOfEnumOP
    })
], ObjectPageSubSectionChartV2.prototype, "visible", void 0);
exports.ObjectPageSubSectionChartV2 = ObjectPageSubSectionChartV2;
class ObjectPageSubSectionAddressV2 {
}
__decorate([
    (0, decorators_1.syncRule)({
        flex: {
            controlId: exports.getSubSectionControlId,
            controlType: () => ux_specification_types_1.ControlType.SubSection,
            createFlexChange: common_1.exportToFlexChange
        },
        generate: common_1.addPatternForBindingChangeOfEnumOP
    })
], ObjectPageSubSectionAddressV2.prototype, "visible", void 0);
exports.ObjectPageSubSectionAddressV2 = ObjectPageSubSectionAddressV2;
class ObjectPageSubSectionContactV2 {
}
__decorate([
    (0, decorators_1.syncRule)({
        flex: {
            controlId: exports.getSubSectionControlId,
            controlType: () => ux_specification_types_1.ControlType.SubSection,
            createFlexChange: common_1.exportToFlexChange
        },
        generate: common_1.addPatternForBindingChangeOfEnumOP
    })
], ObjectPageSubSectionContactV2.prototype, "visible", void 0);
exports.ObjectPageSubSectionContactV2 = ObjectPageSubSectionContactV2;
class ObjectPageSubSectionTableV2 {
}
__decorate([
    (0, decorators_1.syncRule)({
        flex: {
            controlId: exports.getTableSectionControlId,
            controlType: () => ux_specification_types_1.ControlType.SmartTable,
            createFlexChange: common_1.exportToFlexChange
        },
        generate: common_1.addPatternForBindingChangeOfEnumOP
    })
], ObjectPageSubSectionTableV2.prototype, "visible", void 0);
exports.ObjectPageSubSectionTableV2 = ObjectPageSubSectionTableV2;
