// @sap-ux/vocabularies-types/vocabularies/Communication.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.CommunicationAnnotationTypes = exports.CommunicationAnnotationTerms = exports.GenderType = exports.PhoneType = exports.ContactInformationType = exports.KindType = void 0;
// EnumType 
var KindType;
(function (KindType) {
    /**
    A single person or entity
    */
    /**
    undefined
    */
    KindType["individual"] = "Communication.KindType/individual";
    /**
    A group of persons or entities
    */
    /**
    undefined
    */
    KindType["group"] = "Communication.KindType/group";
    /**
    An organization
    */
    /**
    undefined
    */
    KindType["org"] = "Communication.KindType/org";
    /**
    A named geographical place
    */
    /**
    undefined
    */
    KindType["location"] = "Communication.KindType/location";
})(KindType = exports.KindType || (exports.KindType = {}));
// EnumType 
var ContactInformationType;
(function (ContactInformationType) {
    /**
    Related to an individual's work place
    */
    /**
    undefined
    */
    ContactInformationType["work"] = "Communication.ContactInformationType/work";
    /**
    Related to an indivdual's personal life
    */
    /**
    undefined
    */
    ContactInformationType["home"] = "Communication.ContactInformationType/home";
    /**
    Preferred-use contact information
    */
    /**
    undefined
    */
    ContactInformationType["preferred"] = "Communication.ContactInformationType/preferred";
})(ContactInformationType = exports.ContactInformationType || (exports.ContactInformationType = {}));
// EnumType 
var PhoneType;
(function (PhoneType) {
    /**
    Work telephone number
    */
    /**
    undefined
    */
    PhoneType["work"] = "Communication.PhoneType/work";
    /**
    Private telephone number
    */
    /**
    undefined
    */
    PhoneType["home"] = "Communication.PhoneType/home";
    /**
    Preferred-use telephone number
    */
    /**
    undefined
    */
    PhoneType["preferred"] = "Communication.PhoneType/preferred";
    /**
    Voice telephone number
    */
    /**
    undefined
    */
    PhoneType["voice"] = "Communication.PhoneType/voice";
    /**
    Cellular or mobile telephone number
    */
    /**
    undefined
    */
    PhoneType["cell"] = "Communication.PhoneType/cell";
    /**
    Facsimile telephone number
    */
    /**
    undefined
    */
    PhoneType["fax"] = "Communication.PhoneType/fax";
    /**
    Video conferencing telephone number
    */
    /**
    undefined
    */
    PhoneType["video"] = "Communication.PhoneType/video";
})(PhoneType = exports.PhoneType || (exports.PhoneType = {}));
// EnumType 
var GenderType;
(function (GenderType) {
    /**
    male
    */
    /**
    undefined
    */
    GenderType["M"] = "Communication.GenderType/M";
    /**
    female
    */
    /**
    undefined
    */
    GenderType["F"] = "Communication.GenderType/F";
    /**
    other
    */
    /**
    undefined
    */
    GenderType["O"] = "Communication.GenderType/O";
    /**
    not applicable
    */
    /**
    undefined
    */
    GenderType["N"] = "Communication.GenderType/N";
    /**
    unknown
    */
    /**
    undefined
    */
    GenderType["U"] = "Communication.GenderType/U";
})(GenderType = exports.GenderType || (exports.GenderType = {}));
var CommunicationAnnotationTerms;
(function (CommunicationAnnotationTerms) {
    CommunicationAnnotationTerms["Contact"] = "com.sap.vocabularies.Communication.v1.Contact";
    CommunicationAnnotationTerms["Address"] = "com.sap.vocabularies.Communication.v1.Address";
    CommunicationAnnotationTerms["IsEmailAddress"] = "com.sap.vocabularies.Communication.v1.IsEmailAddress";
    CommunicationAnnotationTerms["IsPhoneNumber"] = "com.sap.vocabularies.Communication.v1.IsPhoneNumber";
    CommunicationAnnotationTerms["Event"] = "com.sap.vocabularies.Communication.v1.Event";
    CommunicationAnnotationTerms["Task"] = "com.sap.vocabularies.Communication.v1.Task";
    CommunicationAnnotationTerms["Message"] = "com.sap.vocabularies.Communication.v1.Message";
})(CommunicationAnnotationTerms = exports.CommunicationAnnotationTerms || (exports.CommunicationAnnotationTerms = {}));
var CommunicationAnnotationTypes;
(function (CommunicationAnnotationTypes) {
    CommunicationAnnotationTypes["ContactType"] = "com.sap.vocabularies.Communication.v1.ContactType";
    CommunicationAnnotationTypes["NameType"] = "com.sap.vocabularies.Communication.v1.NameType";
    CommunicationAnnotationTypes["AddressType"] = "com.sap.vocabularies.Communication.v1.AddressType";
    CommunicationAnnotationTypes["PhoneNumberType"] = "com.sap.vocabularies.Communication.v1.PhoneNumberType";
    CommunicationAnnotationTypes["EmailAddressType"] = "com.sap.vocabularies.Communication.v1.EmailAddressType";
    CommunicationAnnotationTypes["GeoDataType"] = "com.sap.vocabularies.Communication.v1.GeoDataType";
    CommunicationAnnotationTypes["UrlType"] = "com.sap.vocabularies.Communication.v1.UrlType";
    CommunicationAnnotationTypes["EventData"] = "com.sap.vocabularies.Communication.v1.EventData";
    CommunicationAnnotationTypes["TaskData"] = "com.sap.vocabularies.Communication.v1.TaskData";
    CommunicationAnnotationTypes["MessageData"] = "com.sap.vocabularies.Communication.v1.MessageData";
})(CommunicationAnnotationTypes = exports.CommunicationAnnotationTypes || (exports.CommunicationAnnotationTypes = {}));
