// @sap/ux-specification/dist/src/sync/v4/export/controls/FilterBar.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.VisualFilter = exports.FilterBar = void 0;
const decorators_1 = require("../../../common/decoration/decorators");
const application_1 = require("../../application");
function exportVisualFilterValueList(manifestSection, configObject, key, manifestKey) {
    if (manifestSection[manifestKey]) {
        manifestSection[manifestKey].valueList = configObject[key];
    }
    else {
        manifestSection[manifestKey] = { valueList: configObject[key] };
    }
}
class FilterBar {
}
__decorate([
    (0, decorators_1.syncRule)({
        manifest: {
            path: (pageName) => (0, application_1.getSettingsPathForPage)(pageName)
        }
    }),
    (0, decorators_1.validity)({
        since: '1.99.0'
    })
], FilterBar.prototype, "hideFilterBar", void 0);
__decorate([
    (0, decorators_1.syncRule)({
        manifest: {
            path: application_1.getSelectionFieldsPath
        }
    }),
    (0, decorators_1.validity)({
        since: '1.93.0'
    })
], FilterBar.prototype, "initialLayout", void 0);
__decorate([
    (0, decorators_1.syncRule)({
        manifest: {
            path: application_1.getSelectionFieldsPath
        }
    }),
    (0, decorators_1.validity)({
        since: '1.93.0'
    })
], FilterBar.prototype, "layout", void 0);
exports.FilterBar = FilterBar;
class VisualFilter {
}
__decorate([
    (0, decorators_1.syncRule)({
        manifest: {
            path: application_1.getFilterFieldPath
        }
    }),
    (0, decorators_1.validity)({
        since: '1.93.0'
    })
], VisualFilter.prototype, "availability", void 0);
__decorate([
    (0, decorators_1.syncRule)({
        manifest: {
            path: application_1.getFilterFieldPath,
            import: (manifestSection) => manifestSection?.['visualFilter']?.['valueList'],
            export: exportVisualFilterValueList,
            key: 'visualFilter'
        }
    }),
    (0, decorators_1.validity)({
        since: '1.93.0'
    })
], VisualFilter.prototype, "visualFilterValueList", void 0);
exports.VisualFilter = VisualFilter;
