// @sap-ux/fe-mockserver-core/dist/mockdata/fileBasedMockData.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.FileBasedMockData = void 0;
const annotation_converter_1 = require("@sap-ux/annotation-converter");
const lodash_clonedeep_1 = __importDefault(require("lodash.clonedeep"));
const common_1 = require("../data/common");
function performSimpleComparison(operator, mockValue, targetLiteral) {
    let isValid = true;
    switch (operator) {
        case 'gt':
            isValid = mockValue > targetLiteral;
            break;
        case 'ge':
            isValid = mockValue >= targetLiteral;
            break;
        case 'lt':
            isValid = mockValue < targetLiteral;
            break;
        case 'le':
            isValid = mockValue <= targetLiteral;
            break;
        case 'ne':
            isValid = mockValue !== targetLiteral;
            break;
        case 'eq':
        default:
            isValid = mockValue === targetLiteral;
            break;
    }
    return isValid;
}
class FileBasedMockData {
    constructor(mockData, entityType, mockDataEntitySet, contextId) {
        this._hierarchyTree = {};
        this._entityType = entityType;
        this._contextId = contextId;
        this._mockDataEntitySet = mockDataEntitySet;
        if (mockData.length === 0 && mockData.__generateMockData) {
            this._mockData = this.generateMockData();
        }
        else {
            this._mockData = (0, lodash_clonedeep_1.default)(mockData);
            if (this._mockData.forEach) {
                this._mockData.forEach((mockLine) => {
                    // We need to ensure that complex types are at least partially created
                    this.validateProperties(mockLine, this._entityType.entityProperties);
                });
            }
        }
    }
    validateProperties(mockEntry, properties) {
        properties.forEach((prop) => {
            var _a, _b;
            if (!prop.nullable &&
                !mockEntry.hasOwnProperty(prop.name) &&
                ((_b = (_a = prop.annotations.Core) === null || _a === void 0 ? void 0 : _a.Computed) === null || _b === void 0 ? void 0 : _b.valueOf()) !== true) {
                mockEntry[prop.name] = this.getDefaultValueFromType(prop.type, prop.targetType, prop.defaultValue);
            }
            else if (mockEntry.hasOwnProperty(prop.name) && (0, annotation_converter_1.isComplexTypeDefinition)(prop.targetType)) {
                // If the property is defined from a complex type we should validate the property of the complex type
                this.validateProperties(mockEntry[prop.name], prop.targetType.properties);
            }
        });
    }
    async addEntry(mockEntry, _odataRequest) {
        this._mockData.push(mockEntry);
    }
    async updateEntry(keyValues, updatedData, _patchData, _odataRequest) {
        const dataIndex = this.getDataIndex(keyValues, _odataRequest);
        this._mockData[dataIndex] = updatedData;
    }
    fetchEntries(keyValues, _odataRequest) {
        const keys = this._entityType.keys;
        return this._mockData.filter((mockData) => {
            return Object.keys(keyValues).every(this.checkKeyValues(mockData, keyValues, keys, _odataRequest));
        });
    }
    hasEntry(keyValues, _odataRequest) {
        return this.getDataIndex(keyValues, _odataRequest) !== -1;
    }
    hasEntries(_odataRequest) {
        return this._mockData.length > 0;
    }
    getAllEntries(_odataRequest, dontClone = false) {
        if (dontClone) {
            return this._mockData;
        }
        return (0, lodash_clonedeep_1.default)(this._mockData);
    }
    getDataIndex(keyValues, _odataRequest) {
        const keys = this._entityType.keys;
        return this._mockData.findIndex((mockData) => {
            return Object.keys(keyValues).every(this.checkKeyValues(mockData, keyValues, keys, _odataRequest));
        });
    }
    checkKeyValues(mockData, keyValues, keys, _odataRequest) {
        return (keyName) => {
            return this._mockDataEntitySet.checkKeyValue(mockData, keyValues, keyName, keys.find((keyProp) => keyProp.name === keyName));
        };
    }
    async removeEntry(keyValues, _odataRequest) {
        const dataIndex = this.getDataIndex(keyValues, _odataRequest);
        if (dataIndex !== -1) {
            this._mockData.splice(dataIndex, 1);
        }
    }
    getDefaultValueFromType(type, complexType, defaultValue) {
        if (complexType) {
            if (complexType._type === 'ComplexType') {
                const outData = {};
                complexType.properties.forEach((subProp) => {
                    outData[subProp.name] = this.getDefaultValueFromType(subProp.type, subProp.targetType, subProp.defaultValue);
                });
                return outData;
            }
            else if (complexType._type === 'TypeDefinition') {
                type = complexType.underlyingType;
            }
        }
        if (defaultValue !== undefined) {
            return defaultValue;
        }
        switch (type) {
            case 'Edm.Int16':
            case 'Edm.Byte':
            case 'Edm.Int32':
            case 'Edm.Int64':
            case 'Edm.Decimal':
                return 0;
            case 'Edm.Boolean':
                return false;
            case 'Edm.DateTimeOffset': {
                const date = new Date();
                return this._mockDataEntitySet.isV4() ? date.toISOString() : '/Date(' + date.getTime() + '+0000)/';
            }
            case 'Edm.Date':
            case 'Edm.DateTime': {
                const date = new Date();
                const dateOut = date.getUTCFullYear() +
                    '-' +
                    ('0' + (date.getUTCMonth() + 1)).slice(-2) +
                    '-' +
                    ('0' + date.getUTCDate()).slice(-2);
                return this._mockDataEntitySet.isV4() ? dateOut : '/Date(' + date.getTime() + '+0000)/';
            }
            case 'Edm.Time':
            case 'Time': {
                const date = new Date();
                // ODataModel expects ISO8601 duration format
                return 'PT' + date.getHours() + 'H' + date.getMinutes() + 'M' + date.getSeconds() + 'S';
            }
            default:
                return '';
        }
    }
    getRandomValueFromType(type, complexType, propertyName, lineIndex) {
        if (complexType) {
            const outData = {};
            if (complexType._type === 'ComplexType') {
                complexType.properties.forEach((subProp) => {
                    outData[subProp.name] = this.getRandomValueFromType(subProp.type, subProp.targetType, subProp.name, lineIndex);
                });
                return outData;
            }
            else if (complexType._type === 'TypeDefinition') {
                type = complexType.underlyingType;
            }
        }
        switch (type) {
            case 'Edm.Int16':
            case 'Edm.Int32':
            case 'Edm.Int64':
                return Math.floor(Math.random() * 10000);
            case 'Edm.String':
                return `${propertyName}_${lineIndex}`;
            case 'Edm.Boolean':
                return Math.random() < 0.5;
            case 'Edm.Byte':
                return Math.floor(Math.random() * 10);
            case 'Edm.Decimal':
                return Math.floor(Math.random() * 100000) / 100;
            case 'Edm.Guid':
                return (0, common_1.uuidv4)();
            case 'Edm.Date':
            case 'Edm.DateTime':
            case 'Edm.DateTimeOffset': {
                const date = new Date();
                date.setFullYear(2000 + Math.floor(Math.random() * 22));
                date.setDate(Math.floor(Math.random() * 30));
                date.setMonth(Math.floor(Math.random() * 12));
                date.setMilliseconds(0);
                if (type === 'Edm.Date') {
                    const dateOut = date.getUTCFullYear() +
                        '-' +
                        ('0' + (date.getUTCMonth() + 1)).slice(-2) +
                        '-' +
                        ('0' + date.getUTCDate()).slice(-2);
                    return this._mockDataEntitySet.isV4() ? dateOut : '/Date(' + date.getTime() + '+0000)/';
                }
                else {
                    return this._mockDataEntitySet.isV4() ? date.toISOString() : '/Date(' + date.getTime() + '+0000)/';
                }
            }
            case 'Edm.Time':
            case 'Time':
                // ODataModel expects ISO8601 duration format
                return ('PT' +
                    Math.floor(Math.random() * 23) +
                    'H' +
                    Math.floor(Math.random() * 59) +
                    'M' +
                    Math.floor(Math.random() * 59) +
                    'S');
            case 'Edm.TimeOfDay':
            case 'Edm.Binary':
            default:
                return '';
        }
    }
    getEmptyObject(_odataRequest) {
        const outObj = {};
        this._entityType.entityProperties.forEach((property) => {
            outObj[property.name] = this.getDefaultValueFromType(property.type, property.targetType, property.defaultValue);
        });
        return outObj;
    }
    getDefaultElement(_odataRequest) {
        if (this._mockData && !Array.isArray(this._mockData)) {
            return this._mockData;
        }
        else if (this._mockData.length >= 1) {
            return (0, lodash_clonedeep_1.default)(this._mockData[0]);
        }
        else {
            return this.getEmptyObject(_odataRequest);
        }
    }
    generateKey(property, lineIndex, mockData = []) {
        const currentMockData = this._mockData || mockData;
        let highestIndex;
        switch (property.type) {
            case 'Edm.Int32':
                highestIndex = 0;
                currentMockData.forEach((mockLine) => {
                    const mockLineIndex = parseInt(mockLine[property.name], 10);
                    highestIndex = Math.max(highestIndex, mockLineIndex);
                });
                return highestIndex + 1;
            case 'Edm.Boolean':
                return Math.random() > 0.5;
            case 'Edm.Guid':
                return (0, common_1.uuidv4)();
            case 'Edm.String':
                if (lineIndex === undefined) {
                    lineIndex = currentMockData.length + 1;
                }
                return `${property.name}_${lineIndex}`;
            default:
                return (0, common_1.generateId)(12);
        }
    }
    generateMockDataLine(iIndex, mockData) {
        const outObj = {};
        this._entityType.entityProperties.forEach((property) => {
            if (property.isKey) {
                outObj[property.name] = this.generateKey(property, iIndex, mockData);
            }
            else {
                outObj[property.name] = this.getRandomValueFromType(property.type, property.targetType, property.name, iIndex);
            }
        });
        this._entityType.navigationProperties.forEach((navigationProperty) => {
            if (navigationProperty.containsTarget) {
                outObj[navigationProperty.name] = [];
            }
        });
        return outObj;
    }
    getParentEntityInterface() {
        return this._mockDataEntitySet.getParentEntityInterface(this._contextId);
    }
    getEntityInterface(entitySetName) {
        return this._mockDataEntitySet.getEntityInterface(entitySetName, this._contextId);
    }
    generateMockData() {
        const mockData = [];
        for (let i = 0; i < 150; i++) {
            mockData.push(this.generateMockDataLine(i, mockData));
        }
        return mockData;
    }
    /**
     * Allow to modify the action data beforehand.
     *
     * @param _actionDefinition
     * @param actionData
     * @param _keys
     * @param _odataRequest
     */
    async onBeforeAction(_actionDefinition, actionData, _keys, _odataRequest) {
        return actionData;
    }
    /**
     * Do something with the action.
     *
     * @param _actionDefinition
     * @param actionData
     * @param _keys
     * @param _odataRequest
     */
    async executeAction(_actionDefinition, actionData, _keys, _odataRequest) {
        return actionData;
    }
    /**
     * Allow to modify the response data.
     *
     * @param _actionDefinition
     * @param _actionData
     * @param _keys
     * @param responseData
     * @param _odataRequest
     */
    async onAfterAction(_actionDefinition, _actionData, _keys, responseData, _odataRequest) {
        return responseData;
    }
    //eslint-disable-next-line
    async onAfterUpdateEntry(_keyValues, _updatedData, _odataRequest) {
        // DO Nothing
    }
    //eslint-disable-next-line
    async onBeforeUpdateEntry(_keyValues, _updatedData, _odataRequest) {
        // DO Nothing
    }
    //eslint-disable-next-line
    hasCustomAggregate(_customAggregateName, _odataRequest) {
        return false;
    }
    //eslint-disable-next-line
    performCustomAggregate(_customAggregateName, _dataToAggregate, _odataRequest) {
        // DO Nothing
    }
    checkSearchQuery(mockValue, searchQuery, _odataRequest) {
        return mockValue === null || mockValue === void 0 ? void 0 : mockValue.toString().includes(searchQuery);
    }
    checkFilterValue(comparisonType, mockValue, literal, operator, _odataRequest) {
        let isValid = true;
        switch (comparisonType) {
            case 'Edm.Boolean':
                isValid = !!mockValue === (literal === 'true');
                break;
            case 'Edm.Byte':
            case 'Edm.Int16':
            case 'Edm.Int32':
            case 'Edm.Int64': {
                const intTestValue = parseInt(literal, 10);
                isValid = performSimpleComparison(operator, mockValue, intTestValue);
                break;
            }
            case 'Edm.Decimal': {
                const decimalTestValue = parseFloat(literal);
                isValid = performSimpleComparison(operator, mockValue, decimalTestValue);
                break;
            }
            case 'Edm.Date':
            case 'Edm.Time':
            case 'Edm.DateTime':
            case 'Edm.DateTimeOffset':
                let targetDateLiteral = literal;
                if (literal && literal.startsWith("datetime'")) {
                    targetDateLiteral = literal.substring(9, literal.length - 1);
                }
                const testValue = new Date(targetDateLiteral).getTime();
                const mockValueDate = new Date(mockValue).getTime();
                isValid = performSimpleComparison(operator, mockValueDate, testValue);
                break;
            case 'Edm.String':
            case 'Edm.Guid':
            default:
                let targetLiteral = literal;
                if (literal && literal.startsWith("guid'")) {
                    targetLiteral = literal.substring(5, literal.length - 1);
                }
                else if (literal && literal.startsWith("'")) {
                    targetLiteral = literal.substring(1, literal.length - 1);
                }
                isValid = performSimpleComparison(operator, mockValue === null || mockValue === void 0 ? void 0 : mockValue.toString(), targetLiteral);
                break;
        }
        return isValid;
    }
    async getReferentialConstraints(_navigationProperty) {
        return undefined;
    }
    buildTree(hierarchyNode, allItems, idNode, parentIdentifier, depth = 0, parentNode) {
        const id = hierarchyNode[idNode];
        const children = allItems[id];
        const resultingChildren = [];
        if (children) {
            for (const child of children) {
                //if (child.$inResultSet === true || child.$inResultSet === undefined) {
                resultingChildren.push(this.buildTree(child, allItems, idNode, parentIdentifier, depth + 1, hierarchyNode));
                //}
            }
        }
        hierarchyNode.$parent = parentNode;
        hierarchyNode.$children = resultingChildren;
        hierarchyNode.$rootDistance = depth;
        return hierarchyNode;
    }
    buildHierarchyTree(hierarchyQualifier, inputSet) {
        var _a, _b;
        //if (!this._hierarchyTree[hierarchyQualifier]) {
        const itemPerParents = {};
        const aggregationAnnotation = (_b = (_a = this._entityType.annotations) === null || _a === void 0 ? void 0 : _a.Aggregation) === null || _b === void 0 ? void 0 : _b[`RecursiveHierarchy#${hierarchyQualifier}`];
        const sourceReference = aggregationAnnotation.ParentNavigationProperty.$target.referentialConstraint[0].sourceProperty;
        inputSet.forEach((item) => {
            if (!itemPerParents[item[sourceReference]]) {
                itemPerParents[item[sourceReference]] = [];
            }
            itemPerParents[item[sourceReference]].push(item);
        });
        this._hierarchyTree[hierarchyQualifier] = itemPerParents;
        return this._hierarchyTree[hierarchyQualifier];
    }
    flattenTree(currentNode, outItems, nodeProperty, distanceFromRootProperty, limitedDescendantCount, matchedDescendantCountProperty, drillStateProperty, depth, toExpand = [], toCollapse = [], toShow = [], toShowAncestors = [], forceExpand = false) {
        var _a, _b;
        let descendantCount = 0;
        const shouldShowAncestor = toShowAncestors.includes(currentNode[nodeProperty]);
        if (currentNode && (depth < 0 || depth > 0 || forceExpand || shouldShowAncestor)) {
            const shouldExpand = toExpand.includes(currentNode[nodeProperty]);
            if (shouldExpand) {
                depth++;
            }
            const shouldShow = toShow.includes(currentNode[nodeProperty]);
            if (shouldShowAncestor && !shouldExpand) {
                forceExpand = true;
            }
            if (shouldShow && !shouldShowAncestor) {
                depth = 1;
            }
            const shouldCollapse = toCollapse.includes(currentNode[nodeProperty]);
            if (shouldCollapse) {
                depth = 1;
            }
            const isLastLevel = depth === 1;
            const isPastLastLevel = depth === 0;
            let wasAdded = false;
            if (!outItems.includes(currentNode) &&
                (!isPastLastLevel || !forceExpand || shouldShow || shouldShowAncestor)) {
                wasAdded = true;
                outItems.push(currentNode);
            }
            if (distanceFromRootProperty) {
                currentNode[distanceFromRootProperty] = currentNode.$rootDistance;
            }
            if (drillStateProperty) {
                if (isLastLevel && ((_a = currentNode.$children) === null || _a === void 0 ? void 0 : _a.length) === 0) {
                    currentNode[drillStateProperty] = 'leaf';
                }
                else if (isLastLevel && !shouldShowAncestor) {
                    currentNode[drillStateProperty] = 'collapsed';
                }
                else {
                    currentNode[drillStateProperty] = 'expanded';
                }
            }
            const children = (_b = currentNode.$children) !== null && _b !== void 0 ? _b : [];
            children.forEach((child) => {
                descendantCount += this.flattenTree(child, outItems, nodeProperty, distanceFromRootProperty, limitedDescendantCount, matchedDescendantCountProperty, drillStateProperty, depth - 1, toExpand, toCollapse, toShow, toShowAncestors, forceExpand);
            });
            if (limitedDescendantCount) {
                currentNode[limitedDescendantCount] = isLastLevel && !shouldShowAncestor ? 0 : descendantCount;
            }
            if (currentNode.$inResultSet && wasAdded) {
                descendantCount++;
            }
            return descendantCount;
        }
        return descendantCount;
    }
    getAncestorsOfNode(currentNode, outItems, nodeProperty, distanceFromRootProperty, matchedDescendantsProperty, matchedProperty, drillStateProperty, depth, matchedChildrenCount, matchedProperties) {
        var _a, _b;
        let descendantCount = 0;
        if (currentNode && (depth < 0 || depth > 0)) {
            const isLastLevel = depth === 1;
            if (outItems.includes(currentNode)) {
                if (matchedDescendantsProperty) {
                    currentNode[matchedDescendantsProperty] += matchedChildrenCount;
                }
            }
            else {
                outItems.push(currentNode);
                if (matchedDescendantsProperty) {
                    currentNode[matchedDescendantsProperty] = matchedChildrenCount;
                }
                if (matchedProperty) {
                    currentNode[matchedProperty] = !!matchedProperties.find((prop) => prop[nodeProperty] === currentNode[nodeProperty]);
                    if (currentNode[matchedProperty]) {
                        matchedChildrenCount++;
                    }
                }
                if (distanceFromRootProperty) {
                    currentNode[distanceFromRootProperty] = currentNode.$rootDistance;
                }
                if (drillStateProperty) {
                    const includedChildrenCount = (_b = (_a = currentNode.$children) === null || _a === void 0 ? void 0 : _a.filter((child) => child.$inResultSet).length) !== null && _b !== void 0 ? _b : 0;
                    if (isLastLevel && includedChildrenCount === 0) {
                        currentNode[drillStateProperty] = 'leaf';
                    }
                    else if (isLastLevel) {
                        currentNode[drillStateProperty] = 'collapsed';
                    }
                    else {
                        currentNode[drillStateProperty] = 'expanded';
                    }
                }
            }
            if (currentNode.$parent) {
                this.getAncestorsOfNode(currentNode.$parent, outItems, nodeProperty, distanceFromRootProperty, matchedDescendantsProperty, matchedProperty, drillStateProperty, depth - 1, matchedChildrenCount, matchedProperties);
            }
            descendantCount++; // also include yourself
            return descendantCount;
        }
        return descendantCount;
    }
    async getTopLevels(data, _parameters, _odataRequest) {
        var _a, _b, _c, _d, _e, _f, _g, _h, _j, _k, _l, _m;
        const hierarchyQualifier = _parameters.HierarchyQualifier.substring(1, _parameters.HierarchyQualifier.length - 1);
        const outItems = [];
        const aggregationAnnotation = (_b = (_a = this._entityType.annotations) === null || _a === void 0 ? void 0 : _a.Aggregation) === null || _b === void 0 ? void 0 : _b[`RecursiveHierarchy#${hierarchyQualifier}`];
        if (aggregationAnnotation) {
            const nodeProperty = aggregationAnnotation.NodeProperty.$target.name;
            const adjustedData = this._mockData.map((item) => {
                const adjustedRowData = data.find((dataItem) => dataItem[nodeProperty] === item[nodeProperty]);
                if (adjustedRowData) {
                    return { ...item, ...adjustedRowData, ...{ $inResultSet: true } };
                }
                else {
                    return { ...item, ...{ $inResultSet: false } };
                }
            });
            const hierarchyNodes = this.buildHierarchyTree(hierarchyQualifier, adjustedData);
            const sourceReference = aggregationAnnotation.ParentNavigationProperty.$target.referentialConstraint[0].sourceProperty;
            // TODO Considering the input set the top level node is not necessarely the root node
            const allRootNodes = adjustedData.filter((node) => {
                const parent = adjustedData.find((parent) => parent[nodeProperty] === node[sourceReference]);
                return !parent || !parent.$inResultSet;
            });
            const depth = parseInt(_parameters.Levels, 10);
            const hierarchyAnnotation = (_d = (_c = this._entityType.annotations) === null || _c === void 0 ? void 0 : _c.Hierarchy) === null || _d === void 0 ? void 0 : _d[`RecursiveHierarchy#${hierarchyQualifier}`];
            let distanceFromRootProperty;
            let matchedDescendantCountProperty;
            let limitedDescendantCountProperty;
            let drillStateProperty;
            if (hierarchyAnnotation) {
                distanceFromRootProperty = (_e = hierarchyAnnotation.DistanceFromRootProperty) === null || _e === void 0 ? void 0 : _e.$target.name;
                limitedDescendantCountProperty = (_f = hierarchyAnnotation.LimitedDescendantCountProperty) === null || _f === void 0 ? void 0 : _f.$target.name;
                matchedDescendantCountProperty = (_g = hierarchyAnnotation.MatchedDescendantCountProperty) === null || _g === void 0 ? void 0 : _g.$target.name;
                drillStateProperty = (_h = hierarchyAnnotation.DrillStateProperty) === null || _h === void 0 ? void 0 : _h.$target.name;
            }
            const toExpand = (_k = (_j = _parameters.Expand) === null || _j === void 0 ? void 0 : _j.map((expand) => expand.substring(1, expand.length - 1))) !== null && _k !== void 0 ? _k : [];
            const toShow = (_m = (_l = _parameters.Show) === null || _l === void 0 ? void 0 : _l.map((collapse) => collapse.substring(1, collapse.length - 1))) !== null && _m !== void 0 ? _m : [];
            const toShowAncestors = [];
            for (const nodeId of toShow) {
                const node = this._mockData.find((node) => node[nodeProperty] === nodeId);
                if (node) {
                    const toShowAncestorsDef = await this.getAncestors(this._mockData, this._mockData, [node], this._entityType, {
                        hierarchyRoot: '',
                        qualifier: hierarchyQualifier,
                        propertyPath: '',
                        maximumDistance: -1,
                        keepStart: false,
                        inputSetTransformations: []
                    }, _odataRequest);
                    toShowAncestorsDef.forEach((ancestor) => {
                        toShowAncestors.push(ancestor[nodeProperty]);
                    });
                }
            }
            allRootNodes.forEach((rootNode) => {
                var _a;
                const hierarchy = this.buildTree(rootNode, hierarchyNodes, nodeProperty, sourceReference, 0, undefined);
                this.flattenTree(hierarchy, outItems, nodeProperty, distanceFromRootProperty, limitedDescendantCountProperty, matchedDescendantCountProperty, drillStateProperty, depth, toExpand, (_a = _parameters.Collapse) === null || _a === void 0 ? void 0 : _a.map((collapse) => collapse.substring(1, collapse.length - 1)), toShow, toShowAncestors);
            });
            let outData = [];
            outItems.forEach((item) => {
                const subTreeData = data.find((dataItem) => dataItem[nodeProperty] === item[nodeProperty]);
                if (subTreeData) {
                    if (matchedDescendantCountProperty &&
                        drillStateProperty &&
                        item[matchedDescendantCountProperty] === 0) {
                        item[drillStateProperty] = 'leaf';
                    }
                    outData.push({ ...subTreeData, ...item });
                }
                else if (toShow.includes(item[nodeProperty]) || toShowAncestors.includes(item[nodeProperty])) {
                    outData.push(item);
                }
            });
            // restrict tree data with skiplocation && skipcontext
            if (_odataRequest.skipLocation) {
                let skipLocation = _odataRequest.skipLocation.split('(')[1].split(')')[0];
                skipLocation = skipLocation.substring(1, skipLocation.length - 1);
                const skipLocationIndex = outData.findIndex((item) => item[nodeProperty] === skipLocation);
                if (skipLocationIndex >= _odataRequest.skipContext) {
                    outData = outData.slice(skipLocationIndex - _odataRequest.skipContext);
                    _odataRequest.addResponseAnnotation('@com.sap.vocabularies.Common.v1.skip', skipLocationIndex - _odataRequest.skipContext);
                }
            }
            return outData;
        }
        return outItems;
    }
    async getDescendants(inputSet, lastFilterTransformationResult, hierarchyFilter, entityType, _parameters, _odataRequest) {
        var _a, _b, _c, _d, _e, _f, _g, _h, _j;
        const aggregationAnnotation = (_b = (_a = this._entityType.annotations) === null || _a === void 0 ? void 0 : _a.Aggregation) === null || _b === void 0 ? void 0 : _b[`RecursiveHierarchy#${_parameters.qualifier}`];
        const hierarchyAnnotation = (_d = (_c = this._entityType.annotations) === null || _c === void 0 ? void 0 : _c.Hierarchy) === null || _d === void 0 ? void 0 : _d[`RecursiveHierarchy#${_parameters.qualifier}`];
        let distanceFromRootProperty;
        let limitedDescendantCountProperty;
        let matchedDescendantCountProperty;
        let drillStateProperty;
        let matchedProperty;
        if (hierarchyAnnotation) {
            distanceFromRootProperty = (_e = hierarchyAnnotation.DistanceFromRootProperty) === null || _e === void 0 ? void 0 : _e.$target.name;
            limitedDescendantCountProperty = (_f = hierarchyAnnotation.LimitedDescendantCountProperty) === null || _f === void 0 ? void 0 : _f.$target.name;
            matchedDescendantCountProperty = (_g = hierarchyAnnotation.MatchedDescendantCountProperty) === null || _g === void 0 ? void 0 : _g.$target.name;
            drillStateProperty = (_h = hierarchyAnnotation.DrillStateProperty) === null || _h === void 0 ? void 0 : _h.$target.name;
            matchedProperty = (_j = hierarchyAnnotation.MatchedProperty) === null || _j === void 0 ? void 0 : _j.$target.name;
        }
        if (aggregationAnnotation) {
            const nodeProperty = aggregationAnnotation.NodeProperty.$target.name;
            const adjustedData = this._mockData.map((item) => {
                const adjustedRowData = hierarchyFilter.find((dataItem) => dataItem[nodeProperty] === item[nodeProperty]);
                if (adjustedRowData) {
                    return { ...item, ...adjustedRowData, ...{ $inResultSet: true } };
                }
                else {
                    return { ...item, ...{ $inResultSet: false } };
                }
            });
            const hierarchyNodes = this.buildHierarchyTree(_parameters.qualifier, adjustedData);
            const sourceReference = aggregationAnnotation.ParentNavigationProperty.$target.referentialConstraint[0].sourceProperty;
            const rootNodes = hierarchyNodes[''];
            rootNodes.forEach((rootNode) => {
                this.buildTree(rootNode, hierarchyNodes, nodeProperty, sourceReference, 0, undefined);
            });
            const subTrees = [];
            hierarchyFilter.forEach((item) => {
                const parentNodeChildren = hierarchyNodes[item[sourceReference]];
                const currentNode = parentNodeChildren.find((node) => node[nodeProperty] === item[nodeProperty]);
                if (_parameters.keepStart) {
                    if (matchedProperty) {
                        // TODO compare with lastFilterTransformationResult
                        currentNode[matchedProperty] = true;
                    }
                    subTrees.push(currentNode);
                }
                currentNode.$children.forEach((child) => {
                    this.flattenTree(child, subTrees, nodeProperty, distanceFromRootProperty, limitedDescendantCountProperty, matchedDescendantCountProperty, drillStateProperty, _parameters.maximumDistance);
                });
            });
            const outData = [];
            inputSet.forEach((item) => {
                const subTreeData = subTrees.find((dataItem) => dataItem[nodeProperty] === item[nodeProperty]);
                if (subTreeData) {
                    if (matchedDescendantCountProperty &&
                        drillStateProperty &&
                        item[matchedDescendantCountProperty] === 0) {
                        subTreeData[drillStateProperty] = 'leaf';
                    }
                    outData.push({ ...item, ...subTreeData });
                }
            });
            return outData;
        }
        else {
            return [];
        }
    }
    async getAncestors(inputSet, lastFilterTransformationResult, limitedHierarchy, entityType, _parameters, _odataRequest) {
        var _a, _b, _c, _d, _e, _f, _g, _h;
        const aggregationAnnotation = (_b = (_a = this._entityType.annotations) === null || _a === void 0 ? void 0 : _a.Aggregation) === null || _b === void 0 ? void 0 : _b[`RecursiveHierarchy#${_parameters.qualifier}`];
        const hierarchyAnnotation = (_d = (_c = this._entityType.annotations) === null || _c === void 0 ? void 0 : _c.Hierarchy) === null || _d === void 0 ? void 0 : _d[`RecursiveHierarchy#${_parameters.qualifier}`];
        let distanceFromRootProperty;
        let matchedDescendantCountProperty;
        let matchedProperty;
        let drillStateProperty;
        if (hierarchyAnnotation) {
            distanceFromRootProperty = (_e = hierarchyAnnotation.DistanceFromRootProperty) === null || _e === void 0 ? void 0 : _e.$target.name;
            drillStateProperty = (_f = hierarchyAnnotation.DrillStateProperty) === null || _f === void 0 ? void 0 : _f.$target.name;
            matchedDescendantCountProperty = (_g = hierarchyAnnotation.MatchedDescendantCountProperty) === null || _g === void 0 ? void 0 : _g.$target.name;
            matchedProperty = (_h = hierarchyAnnotation.MatchedProperty) === null || _h === void 0 ? void 0 : _h.$target.name;
        }
        if (aggregationAnnotation) {
            const nodeProperty = aggregationAnnotation.NodeProperty.$target.name;
            const sourceReference = aggregationAnnotation.ParentNavigationProperty.$target.referentialConstraint[0].sourceProperty;
            const adjustedData = this._mockData.map((item) => {
                const adjustedRowData = limitedHierarchy.find((dataItem) => dataItem[nodeProperty] === item[nodeProperty]);
                if (adjustedRowData) {
                    return { ...item, ...adjustedRowData, ...{ $inResultSet: true } };
                }
                else {
                    return { ...item, ...{ $inResultSet: false } };
                }
            });
            const hierarchyNodes = this.buildHierarchyTree(_parameters.qualifier, adjustedData);
            const rootNodes = hierarchyNodes[''];
            rootNodes.forEach((rootNode) => {
                this.buildTree(rootNode, hierarchyNodes, nodeProperty, sourceReference, 0, undefined);
            });
            const ancestors = [];
            limitedHierarchy.forEach((item) => {
                const parentNodeChildren = hierarchyNodes[item[sourceReference]];
                const currentNode = parentNodeChildren.find((node) => node[nodeProperty] === item[nodeProperty]);
                if (_parameters.keepStart) {
                    this.getAncestorsOfNode(currentNode, ancestors, nodeProperty, distanceFromRootProperty, matchedDescendantCountProperty, matchedProperty, drillStateProperty, _parameters.maximumDistance - 1, 0, lastFilterTransformationResult);
                }
                else if (currentNode && currentNode.$parent) {
                    this.getAncestorsOfNode(currentNode.$parent, ancestors, nodeProperty, distanceFromRootProperty, matchedDescendantCountProperty, matchedProperty, drillStateProperty, _parameters.maximumDistance - 1, 1, lastFilterTransformationResult);
                }
            });
            const outData = [];
            inputSet.forEach((item) => {
                const subTreeData = ancestors.find((dataItem) => dataItem[nodeProperty] === item[nodeProperty]);
                if (subTreeData) {
                    outData.push({ ...item, ...subTreeData });
                }
            });
            return outData;
        }
        else {
            return [];
        }
    }
}
exports.FileBasedMockData = FileBasedMockData;
