// @sap-ux/vocabularies-types/vocabularies/Aggregation.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.AggregationAnnotationTypes = exports.AggregationAnnotationTerms = exports.RollupType = void 0;
// EnumType 
/**
  The number of `rollup` or `rollupall` operators allowed in a `groupby` transformation
*/
var RollupType;
(function (RollupType) {
    /**
    No support for `rollup` or `rollupall`
    */
    /**
    undefined
    */
    RollupType["None"] = "Aggregation.RollupType/None";
    /**
    Only one `rollup` or `rollupall` operator per `groupby`
    */
    /**
    undefined
    */
    RollupType["SingleHierarchy"] = "Aggregation.RollupType/SingleHierarchy";
    /**
    Full support for `rollup` and `rollupall`
    */
    /**
    undefined
    */
    RollupType["MultipleHierarchies"] = "Aggregation.RollupType/MultipleHierarchies";
})(RollupType = exports.RollupType || (exports.RollupType = {}));
var AggregationAnnotationTerms;
(function (AggregationAnnotationTerms) {
    AggregationAnnotationTerms["ApplySupported"] = "Org.OData.Aggregation.V1.ApplySupported";
    AggregationAnnotationTerms["ApplySupportedDefaults"] = "Org.OData.Aggregation.V1.ApplySupportedDefaults";
    AggregationAnnotationTerms["Groupable"] = "Org.OData.Aggregation.V1.Groupable";
    AggregationAnnotationTerms["Aggregatable"] = "Org.OData.Aggregation.V1.Aggregatable";
    AggregationAnnotationTerms["CustomAggregate"] = "Org.OData.Aggregation.V1.CustomAggregate";
    AggregationAnnotationTerms["ContextDefiningProperties"] = "Org.OData.Aggregation.V1.ContextDefiningProperties";
    AggregationAnnotationTerms["LeveledHierarchy"] = "Org.OData.Aggregation.V1.LeveledHierarchy";
    AggregationAnnotationTerms["RecursiveHierarchy"] = "Org.OData.Aggregation.V1.RecursiveHierarchy";
    AggregationAnnotationTerms["AvailableOnAggregates"] = "Org.OData.Aggregation.V1.AvailableOnAggregates";
})(AggregationAnnotationTerms = exports.AggregationAnnotationTerms || (exports.AggregationAnnotationTerms = {}));
var AggregationAnnotationTypes;
(function (AggregationAnnotationTypes) {
    AggregationAnnotationTypes["ApplySupportedBase"] = "Org.OData.Aggregation.V1.ApplySupportedBase";
    AggregationAnnotationTypes["ApplySupportedType"] = "Org.OData.Aggregation.V1.ApplySupportedType";
    AggregationAnnotationTypes["AggregatablePropertyType"] = "Org.OData.Aggregation.V1.AggregatablePropertyType";
    AggregationAnnotationTypes["RecursiveHierarchyType"] = "Org.OData.Aggregation.V1.RecursiveHierarchyType";
    AggregationAnnotationTypes["AvailableOnAggregatesType"] = "Org.OData.Aggregation.V1.AvailableOnAggregatesType";
    AggregationAnnotationTypes["NavigationPropertyAggregationCapabilities"] = "Org.OData.Aggregation.V1.NavigationPropertyAggregationCapabilities";
    AggregationAnnotationTypes["CustomAggregateType"] = "Org.OData.Aggregation.V1.CustomAggregateType";
})(AggregationAnnotationTypes = exports.AggregationAnnotationTypes || (exports.AggregationAnnotationTypes = {}));
