// @sap-ux/fe-mockserver-core/dist/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const path = __importStar(require("path"));
const router_1 = __importDefault(require("router"));
const middleware_1 = require("./middleware");
__exportStar(require("./api"), exports);
class FEMockserver {
    constructor(configuration) {
        this.configuration = configuration;
        this.mainRouter = new router_1.default();
        this.isReady = this.initialize();
    }
    async initialize() {
        var _a, _b;
        const FileLoaderClass = this.configuration.fileLoader || (await Promise.resolve().then(() => __importStar(require('./plugins/fileSystemLoader')))).default;
        this.fileLoader = new FileLoaderClass();
        const MetadataProviderClass = (await this.fileLoader.loadJS(((_a = this.configuration.metadataProcessor) === null || _a === void 0 ? void 0 : _a.name) || path.resolve(__dirname, './plugins/metadataProvider'))).default;
        this.metadataProvider = new MetadataProviderClass(this.fileLoader, (_b = this.configuration.metadataProcessor) === null || _b === void 0 ? void 0 : _b.options);
        await (0, middleware_1.createMockMiddleware)(this.configuration, this.mainRouter, this.fileLoader, this.metadataProvider);
    }
    getRouter() {
        return this.mainRouter;
    }
}
exports.default = FEMockserver;
