// @sap-ux/fe-mockserver-core/dist/router/catalogServiceRouter.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.catalogServiceRouter = void 0;
const router_1 = __importDefault(require("router"));
/**
 * Create a service mocking the catalog service from the ABAP backend.
 *
 * @param servicesConfig the service configuration.
 * @param version the version of the services
 * @returns a sub router
 */
function catalogServiceRouter(servicesConfig, version = 2) {
    const router = new router_1.default();
    router.get('/ServiceCollection', (_req, res) => {
        res.setHeader('content-type', 'application/json');
        res.write(JSON.stringify({
            d: {
                results: servicesConfig.map((serviceConfig) => {
                    return {
                        ID: serviceConfig._internalName
                    };
                })
            }
        }));
        res.end();
    });
    router.get('/ServiceCollection\\(*', (_req, res) => {
        res.setHeader('content-type', 'application/json');
        res.write(JSON.stringify({
            d: {
                results: servicesConfig.map((serviceConfig) => {
                    return {
                        TechnicalName: serviceConfig._internalName,
                        Version: version
                    };
                })
            }
        }));
        res.end();
    });
    return router;
}
exports.catalogServiceRouter = catalogServiceRouter;
